pkg/util/arith.go

```diff
@@ -33,3 +33,31 @@ func Min[T constraints.Ordered](x, y T) T {
 		return y
 	}
 }
+
+// AtomicInt represents the shared interface provided by various atomic.<NAME> integers
+//
+// This interface type is primarily used by AtomicMax.
+type AtomicInt[I any] interface {
+	Add(delta I) (new I)                      //nolint:predeclared // same var names as methods
+	CompareAndSwap(old, new I) (swapped bool) //nolint:predeclared // same var names as methods
+	Load() I
+	Store(val I)
+	Swap(new I) (old I) //nolint:predeclared // same var names as methods
+}
+
+// AtomicMax atomically sets a to the maximum of *a and i, returning the old value at a.
+//
+// Eventually, one would *hope* that go could add atomic maximum (and minimum), so we don't need
+// this function (there are some ISAs that support such instructions). In the meantime though, this
+// tends to be pretty useful - event though it's not wait-free.
+func AtomicMax[A AtomicInt[I], I constraints.Integer](a A, i I) I {
+	for {
```

Comment:
I _think_ that this is how `atomic.Value.CompareAndSwap()` is implemented, that not withstanding, I have this feeling that if you get into this situation you probably want a different sort of operation (e.g. just a mutex protecting an int, then you can let the runtime be smart (potentially) about letting the goroutine calling this function sleep until it can be woken up logically.)

Reply:
> I think that this is how atomic.Value.CompareAndSwap() is implemented

_Kind of_, but not quite imo. It's [mostly pretty readable](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/sync/atomic/value.go;l=135), though!